stan/math/prim/scal/meta/require_generics.hpp

```diff
@@ -0,0 +1,769 @@
+#ifndef STAN_MATH_PRIM_SCAL_META_REQUIRE_GENERICS_HPP
+#define STAN_MATH_PRIM_SCAL_META_REQUIRE_GENERICS_HPP
+
+#include <stan/math/prim/scal/meta/bool_constant.hpp>
+#include <stan/math/prim/scal/meta/is_eigen.hpp>
+#include <stan/math/prim/scal/meta/is_fvar.hpp>
+#include <stan/math/prim/scal/meta/is_var.hpp>
+#include <stan/math/prim/scal/meta/is_var_or_arithmetic.hpp>
+#include <stan/math/prim/scal/meta/is_vector.hpp>
+#include <stan/math/prim/scal/meta/is_vector_like.hpp>
+#include <stan/math/prim/scal/meta/scalar_type.hpp>
+#include <stan/math/prim/scal/meta/value_type.hpp>
+#include <stan/math/prim/scal/meta/conjunction.hpp>
+#include <stan/math/prim/scal/meta/disjunction.hpp>
+
+#include <type_traits>
+
+namespace stan {
+
+/**
+ * If condition is true, template is enabled
+ */
+template <class Check>
+using require_t = std::enable_if_t<Check::value>;
+
+/**
+ * If condition is false, template is disabled
+ */
+template <typename Check>
+using require_not_t = std::enable_if_t<!Check::value>;
+
+/**
+ * If all conditions are true, template is enabled
+ * Returns a type void if all conditions are true and otherwise fails.
+ */
+template <class... Checks>
+using require_all_t = std::enable_if_t<math::conjunction<Checks...>::value>;
+
+/**
+ * If any condition is true, template is enabled.
+ *
+ * Returns a type void if any of the conditions are true and otherwise fails.
+ */
+template <class... Checks>
+using require_any_t = std::enable_if_t<math::disjunction<Checks...>::value>;
+
+/**
+ * If all conditions are false, template is enabled.
+ *
+ * Returns a type void if all of the conditions are false.
+ */
+template <class... Checks>
+using require_all_not_t
+    = std::enable_if_t<!math::conjunction<Checks...>::value>;
+
+/**
+ * If any condition is false, template is enabled.
+ *
+ * Returns a type void if any of the conditions are false.
+ */
+template <class... Checks>
+using require_any_not_t
+    = std::enable_if_t<!math::disjunction<Checks...>::value>;
+
+// Enablers for two types of the same value
+template <typename T, typename S>
+using require_same_t
+    = require_t<std::is_same<std::decay_t<T>, std::decay_t<S>>>;
+
+template <typename T, typename S>
+using require_not_same_t
+    = require_not_t<std::is_same<std::decay_t<T>, std::decay_t<S>>>;
+
+template <typename T, typename... Types>
+using require_all_same_t
+    = require_all_t<std::is_same<std::decay_t<T>, std::decay_t<Types>>...>;
+
+template <typename T, typename... Types>
+using require_all_not_same_t
+    = require_all_not_t<std::is_same<std::decay_t<T>, std::decay_t<Types>>...>;
```

Comment:
I think I flat out need to rethink how same works here. Reading it now it checks that the `Types` match with just `T`, but it really should check that each type is the same. If i figure that out I think this will make more sense / I can have an `any` for same as well

Reply:
Actually I think just changing it to `any` makes sense (also i don't feel like writing out the more complicated thing which I'm not sure we would use that much)